pkg/sos/objectstorage_test.go

```diff
@@ -0,0 +1,197 @@
+package sos
+
+import (
+	"context"
+	"github.com/exoscale/egoscale/v2/oapi"
+	"github.com/stretchr/testify/assert"
+	db "github.com/vshn/exoscale-metrics-collector/pkg/database"
+	exoscalev1 "github.com/vshn/provider-exoscale/apis/exoscale/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"testing"
+	"time"
+)
+
+func TestObjectStorage_GetBillingDate(t *testing.T) {
+	t.Run("GivenContext_WhenGetBillingDate_ThenReturnYesterdayDate", func(t *testing.T) {
+		// Given
+		ctx := context.Background()
+		utc, _ := time.LoadLocation("UTC")
+		now := time.Now().In(utc)
+		expected := time.Date(now.Year(), now.Month(), time.Now().Day()-1, 6, 0, 0, 0, now.Location())
+
+		//When
+		o := ObjectStorage{}
+		err := o.getBillingDate(ctx)
+
+		// Then
+		assert.NoError(t, err)
+		assert.Equal(t, o.billingDate, expected)
```

Comment:
This makes the test somewhat hard to understand and very dynamic based on the current time.
I'd rather use a fixed clock here where we can simulate the exact time(s) and verify the expected results.

To achieve a fixed clock, a simple way would be following:
```go
// objectstorage.go
var nowFn = time.Now // var to a function

func getBillingDate()... {
  now := nowFn() // instead of time.Now()
  ...
}
```
and in our test case (prosa, doesn't compile):
```suggestion
		// Given
		ctx := context.Background()
		utc, _ := time.LoadLocation("UTC")
		nowFn = func() time.Time {
		  return time.Date(2022, time.February, 23, ...).Location(utc) // return our fixed time here
		}
		defer func() {nowFn = time.Now}
		expected := time.Date(2022, time.February, 22, etc...) // fixed expected time

		//When
		o := ObjectStorage{}
		err := o.getBillingDate(ctx)

		// Then
		assert.NoError(t, err)
		assert.Equal(t, o.billingDate, expected)
```

Reply:
I really wouldn't say that the version you proposed is more readable... why would you even use a defer in this case? it gets more complicated...

Reply:
My issue is that `now()` is a dynamic value, it changes every time you execute it, even in unit tests. In extreme cases, unit tests may start failing in a few months when switching to summertime or back, in a random unrelated PR for example.
When we have a fixed clock, we can simulate different timestamps and test with different times and timezones, which is not case with the current test implementation.

Unfortunately in Go it's not so easy to simulate a fixed clock, that's why having a function pointer is needed here. We need `defer` to reset the global variable `nowFn` to not influence potentially other test cases that make use of now().

It doesn't make the test **code** easier, true, but it allows us to clearly define the input and expected output and that's what makes the test **case** understandable.

Reply:
> My issue is that now() is a dynamic value, it changes every time you execute it, even in unit tests. In extreme cases, unit tests may start failing in a few months when switching to summer time or back, in a random unrelated PR for example.

I also don't like that it's dynamic but I also don't like that I have to change the production code because of a test. 
If you consider a bit more what we test then you can see it cannot fail, for one reason is that only the date has the dynamic part and second, it's the same code to generate the real value of the billing date. I just want to make sure the billing date is not changed by a developer by mistake in the future. It should not even fail when we move to summer time as the time is fixed.

> When we have a fixed clock, we can simulate different timestamps and test with different times and timezones, which is not case with the current test implementation.

Here I'm not interested to test different times and timezone. The test only has to check that the billing date is not changed by someone by mistake.

> Unfortunately in Go it's not so easy to simulate a fixed clock, that's why having a function pointer is needed here. We need defer to reset the global variable nowFn to not influence potentially other test cases that make use of now().

This is one of the reason I avoided to complicate the test and change the production code in the first place for my test case.

Reply:
I don't quite understand your arguments, but it's ok since I won't be maintaining this code in future :)